R/collector.R

```diff
@@ -149,7 +149,7 @@ env_clone_lazy <- function(env) {
   # https://github.com/cynkra/collector/issues/2
   rm(list = nms, envir = clone)
   for (nm in nms) {
-    env_bind_lazy(clone, !!nm := !!env[[nm]])
+    env_bind_lazy(clone, !!nm := env[[!!nm]])
```

Comment:
Before this modification we were attempting to bind lazily a literal and it appears not to be possible, so the strategy to drop data that isn't used wasn't actually triggered AFAICT.
Now we use the execution environment of the function but it should be fine because after the cleaning step it's not needed and should be garbage collected, if `env_bind_lazy()` works correctly that it.

I'm a bit worried about https://github.com/r-lib/rlang/issues/1703 though, because of it we still might keep extra data ultimately, I wonder if we should not switch to delayedAssign. Or at least check that the cleanup step is working.

Reply:
I think this is fine, this is expanded to a series of calls like `env_bind_lazy(clone, "<my-name>" = env[["<my-name>"]])` . Previously, the expansion was like `env_bind_lazy(clone, "<my-name>" = <my-value>)`, which also works, except for the corner case of symbols and expressions.

What am I missing?

Reply:
As I understand it: The previous strategy was attempting to bind lazily a literal, this creates a standard binding. This means when we check what was used or not we find that everything was "used" (non lazy), and we save too much data. 
Now with the new strategy this might work better because we use an expression in env_bind_lazy and not a literal, so it should create an actual lazy binding, except if it doesn't due to the bug (IMO) described in the ticket above.

Reply:
Any way let's merge this, this improves the previous behavior

Reply:
I don't think the last approach created a standard binding. It looks to me more like a lazy binding to a constant. The current code creates a lazy binding to an expression that is a simple environment lookup. Both are equivalent in terms of being able to detect if a binding was used, but only the latter avoids premature unquoting.

``` r
library(rlang)

a <- 5
a_name <- "a"
# Previous behavior
expr(env_bind_lazy(.GlobalEnv, b = !!sym(a_name)))
#> env_bind_lazy(.GlobalEnv, b = a)
env_bind_lazy(.GlobalEnv, b = !!sym(a_name))
b
#> [1] 5

# Current behavior
expr(env_bind_lazy(.GlobalEnv, c = sym(!!a_name)))
#> env_bind_lazy(.GlobalEnv, c = sym("a"))
env_bind_lazy(.GlobalEnv, c = sym(!!a_name))
c
#> a
```

<sup>Created on 2024-04-30 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

Reply:
~~Better~~ More verbose reprex:

``` r
library(rlang)

a <- 5
a_name <- "a"
# Previous behavior
expr(env_bind_lazy(.GlobalEnv, b = {
  message("b triggered")
  !!sym(a_name)
}))
#> env_bind_lazy(.GlobalEnv, b = {
#>     message("b triggered")
#>     a
#> })
env_bind_lazy(.GlobalEnv, b = {
  message("b triggered")
  !!sym(a_name)
})
b
#> b triggered
#> [1] 5

# Current behavior
expr(env_bind_lazy(.GlobalEnv, c = {
  message("c triggered")
  sym(!!a_name)
}))
#> env_bind_lazy(.GlobalEnv, c = {
#>     message("c triggered")
#>     sym("a")
#> })
env_bind_lazy(.GlobalEnv, c = {
  message("c triggered")
  sym(!!a_name)
})
c
#> c triggered
#> a
```

<sup>Created on 2024-04-30 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

Reply:
I thought that we couldn't set a lazy binding to a literal (in your reprexes these are symbols it's different). But I see that it's weirder than this, the behavior is erratic: 

```
env <- new.env()
env$a <- 1
nm <- "a"
rlang::env_bind_lazy(.GlobalEnv, A = !!env[[nm]])
rlang::env_binding_are_lazy(.GlobalEnv, "A") 
```

<img width="350" alt="Screenshot 2024-04-30 at 12 09 56" src="https://github.com/cynkra/collector/assets/18351714/90a055b2-6895-44c5-bd7d-16e4d154ff89">

Reply:
`delayedAssign()` is also erratic, something like : 

```
delayedAssign("A", 1)
rlang::env_binding_are_lazy(.GlobalEnv, "A") 
```

might sometimes return FALSE, sometimes TRUE

Reply:
Have you tested on the command line in the terminal, without RStudio?

Reply:
I now did, and I haven't been able to get a weird behavior in the R Gui

Reply:
Yeah, the environment pane. All well for me, let's revisit when new problems occur.